contrib/replicate-build/template/predict.py

```diff
@@ -0,0 +1,135 @@
+import asyncio
+import json
+import logging
+import os
+import sys
+import torch
+import transformers
+from queue import Queue
+from typing import Iterator
+
+import cog
+import lmql
+from lmql.models.lmtp.lmtp_scheduler import Scheduler, TokenSession
+from lmql.models.lmtp.lmtp_async import LMTPAsyncTransport
+
+logging.basicConfig()
+logging.getLogger("asyncio").setLevel(logging.WARNING)
+
+os.environ['TRANSFORMERS_OFFLINE']='1'
+
+class Predictor(cog.BasePredictor):
+    def setup(self):
+        """Run initial setup to enable faster predict calls later"""
+        self.model: str = "checkpoints"
+        self.model_args: dict = {"device_map": "auto"}
+
+        if os.path.exists('config.json'):
+            self.model_args.update(json.load(open('config.json', 'r')))
+
+        self.transformers_config = json.load(open('checkpoints/config.json'))
+        self.torch_dtype_name = self.transformers_config.get('torch_dtype')
+        if self.torch_dtype_name is not None:
+            self.torch_dtype = getattr(torch, self.torch_dtype_name)
+        else:
+            self.torch_dtype = None
+
+        if self.torch_dtype is not None:
+            if self.model_args.get('load_in_4bit'):
+                self.model_args['quantization_config'] = transformers.BitsAndBytesConfig(
+                    load_in_4bit = True,
+                    bnb_4bit_compute_dtype = self.torch_dtype,
+                    bnb_4bit_quant_type = 'nf4',
+                    bnb_4bit_use_double_quant=True,
+                )
+
+        # instantiating a scheduler puts it in the registry; we don't _really_ need to keep a reference here
+        self._scheduler = Scheduler.instance(
+            self.model, self.model_args, user=None, sync=False
+        )
+
+    # based on lmtp_inference_task but not using gather
+    async def lmtp_inference_stream(self, ops_batch, loop=None):
+        if loop is None:
+            loop = asyncio.get_event_loop()
+
+        token_queue = asyncio.Queue()
+        transport = LMTPAsyncTransport(token_queue)
+        session = TokenSession(transport, self.model_args, static=True, longrunning=True)
+
+        next_local_id = 0
+        orig_ids = {}     # tracks all ids seen, even ones that are done
```

Comment:
How long will an inference server be running and will this set always grow during that time? Could this become a problem with high uptime?

Reply:
Since we tear down the event loop between predictions, this should be very short-lived.